src/PETScNonlinearSolvers.jl

```diff
@@ -171,16 +159,16 @@ function Algebra.solve!(x::T,
                         cache::PETScNonlinearSolverCache{<:T}) where T <: AbstractVector
 
   @assert cache.op === op
-  @check_error_code PETSC.SNESSolve(nls.snes[],C_NULL,cache.x_petsc_vec.vec[])
+  @check_error_code PETSC.SNESSolve(cache.snes[],C_NULL,cache.x_petsc_vec.vec[])
   _copy_and_exchange!(x,cache.x_petsc_vec)
   cache
 end
 
-function Algebra.solve!(x::AbstractVector,nls::PETScNonlinearSolver,op::NonlinearOperator)
+function Algebra.solve!(x::AbstractVector,nls::PETScNonlinearSolver,op::NonlinearOperator,::Nothing)
   cache=_setup_cache(x,nls,op)
   _set_petsc_residual_function!(nls,cache)
   _set_petsc_jacobian_function!(nls,cache)
-  @check_error_code PETSC.SNESSolve(nls.snes[],C_NULL,cache.x_petsc_vec.vec[])
+  @check_error_code PETSC.SNESSolve(cache.snes[],C_NULL,cache.x_petsc_vec.vec[])
   _copy_and_exchange!(x,cache.x_petsc_vec)
```

Comment:
It is not needed to call `exchange!`. It will be already called when building the FEFunction. 

`copy!(x,cache.x_petsc_vec)` should be OK

Reply:
Done 9b21f583d4a59c98192ec1a6b65c44b2ca257916